binary_search_tree/tree.py

```diff
@@ -7,52 +7,142 @@ def __init__(self, key, val = None):
+    # Space Complexity: O(1)
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+
+        if self.root == None: 
+            self.root = new_node
+            return self.root
+
+        self.add_helper(self.root, new_node)
+
+    def find_helper(self, current_node, key):
+        if current_node == None: 
+            return None 
 
-    # Time Complexity: 
-    # Space Complexity: 
+        if current_node.key == key: 
+            return current_node.value
+
+        if current_node.key > key: 
+            return self.find_helper(current_node.left, key)
+        else: 
+            return self.find_helper(current_node.right, key)
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find(self, key):
-        pass
+        if self.root == None: 
+            return None
+        
+        return self.find_helper(self.root, key)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    def inorder_helper(self, current_node, items):
+        if current_node != None: 
+            self.inorder_helper(current_node.left, items)
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.inorder_helper(current_node.right, items)
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def inorder(self):
-        pass
+        items = []
+
+        self.inorder_helper(self.root, items)
+
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def preorder_helper(self, current_node, items):
+        if current_node != None: 
+            items.append({"key": current_node.key, "value": current_node.value})
+            self.preorder_helper(current_node.left, items)
+            self.preorder_helper(current_node.right, items)
+
+    # Time Complexity: O(n)
+    # Space Complexity:  O(n)   
     def preorder(self):
-        pass
+        items = []
+
+        self.preorder_helper(self.root, items)
+
+        return items
 
-    # Time Complexity: 
-    # Space Complexity:     
+    def postorder_helper(self, current_node, items):
+        if current_node != None: 
+            self.postorder_helper(current_node.left, items)
+            self.postorder_helper(current_node.right, items)    
+            items.append({"key": current_node.key, "value": current_node.value})  
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)  
     def postorder(self):
-        pass
+        items = []
+
+        self.postorder_helper(self.root, items)
+
+        return items
+
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)   
+            
+    def height_helper(self, current_node):
+        if current_node == None: 
+            return 0
+
+        left_tree_height = self.height_helper(current_node.left)
+        right_tree_height = self.height_helper(current_node.right)
+        return max(left_tree_height, right_tree_height) + 1
 
-    # Time Complexity: 
-    # Space Complexity:     
     def height(self):
```

Comment:
✨Nice! Time and space complexity would indeed be O(n) if the tree is unbalanced! For a balanced tree, time and space complexity would be O(log n) because your recursive calls only go through half the nodes!